### ISSUE TITLE
[Impeller] [Flutter 3.10]: Codec failed to produce an image

### ISSUE THREAD

**Issue by sinahamedi67** [flutter/flutter#126768]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

1. Create a new flutter project or open an existing one
2. Try showing an image using `Image.Network` or `Image.asset` on iOS simulator

It happens on different ios simulator with any images from network or assets. Using cached_network_image package also produces the same result. 
This bug produced after upgrading from Flutter 3.3.0 to 3.10 

**Update:** It works fine until 3.7.12

### Expected results

Showing the image successfully

### Actual results

Following exception throws:

```
======== Exception caught by image resource service ================================================
The following _Exception was thrown resolving an image frame:
Exception: Codec failed to produce an image, possibly due to invalid image data.

When the exception was thrown, this was the stack: 
Image provider: NetworkImage("https://picsum.photos/200/300", scale: 1.0)
Image key: NetworkImage("https://picsum.photos/200/300", scale: 1.0)
====================================================================================================
```

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Scaffold(
        body: Center(
          child: Image.network(
            'https://picsum.photos/200/300',
          ),
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

![image](https://github.com/flutter/flutter/assets/78305390/40f45b15-17da-4ec5-a311-d53fd1966f8f)

</details>

### Logs

<details><summary>Logs</summary>

```
[ +816 ms] com.example.images: 64717
[        ] Waiting for VM Service port to be available...
[ +899 ms] VM Service URL on device: http://127.0.0.1:55892/U1waQeMPT4c=/
[   +6 ms] Caching compiled dill
[ +255 ms] Connecting to service protocol: http://127.0.0.1:55892/U1waQeMPT4c=/
[ +347 ms] Launching a Dart Developer Service (DDS) instance at http://127.0.0.1:0, connecting to VM service at http://127.0.0.1:55892/U1waQeMPT4c=/.
[ +256 ms] DDS is listening at http://127.0.0.1:55895/LMq3joqnDbk=/.
[  +95 ms] Successfully connected to service protocol: http://127.0.0.1:55892/U1waQeMPT4c=/
[  +91 ms] DevFS: Creating new filesystem on the device (null)
[  +95 ms] DevFS: Created new filesystem on the device
(file:///Users/flutter/Library/Developer/CoreSimulator/Devices/D4909C3C-EA95-4F22-A52C-62D212410360/data/Containers/Data/Application/A70E296E-62C3-420B-849C-DF75398898BA/t
mp/imagesgj5pit/images/)
[   +2 ms] Updating assets
[ +356 ms] Manifest contained wildcard assets. Inserting missing file into build graph to force rerun. for more information see #56466.
[   +8 ms] Syncing files to device iPhone 13 Pro...
[   +3 ms] Compiling dart to kernel with 0 updated files
[        ] Processing bundle.
[   +1 ms] <- recompile package:images/main.dart 7793d8de-8aba-442e-9871-f95e67dd6cd1
[        ] <- 7793d8de-8aba-442e-9871-f95e67dd6cd1
[  +11 ms] Bundle processing done.
[ +154 ms] Updating files.
[        ] DevFS: Sync finished
[   +2 ms] Syncing files to device iPhone 13 Pro... (completed in 173ms)
[   +1 ms] Synced 0.0MB.
[   +2 ms] <- accept
[  +18 ms] Connected to _flutterView/0x7ff36102ca20.
[ +206 ms] Flutter run key commands.
[  +15 ms] r Hot reload. 🔥🔥🔥
[   +2 ms] R Hot restart.
[        ] h List all available interactive commands.
[  +91 ms] d Detach (terminate "flutter run" but leave application running).
[        ] c Clear the screen
[        ] q Quit (terminate the application on the device).
[   +1 ms] A Dart VM Service on iPhone 13 Pro is available at: http://127.0.0.1:55895/LMq3joqnDbk=/
[  +53 ms] The Flutter DevTools debugger and profiler on iPhone 13 Pro is available at: http://127.0.0.1:9100?uri=http://127.0.0.1:55895/LMq3joqnDbk=/
[+1704 ms] ══╡ EXCEPTION CAUGHT BY IMAGE RESOURCE SERVICE ╞════════════════════════════════════════════════════
                    The following _Exception was thrown resolving an image frame:
                    Exception: Codec failed to produce an image, possibly due to invalid image data.
                    
                    When the exception was thrown, this was the stack
                    
                    Image provider: NetworkImage("https://picsum.photos/200/300", scale: 1.0)
                    Image key: NetworkImage("https://picsum.photos/200/300", scale: 1.0)
                    ════════════════════════════════════════════════════════════════════════════════════════════════════
```
Complete `flutter run --verbose`:
https://pastebin.com/w1gyiz9M

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.10.0, on macOS 12.6 21G115 darwin-x64, locale en-US)
    • Flutter version 3.10.0 on channel stable at /Users/flutter/Downloads/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 84a1e904f4 (5 days ago), 2023-05-09 07:41:44 -0700
    • Engine revision d44b5a94c9
    • Dart version 3.0.0
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/flutter/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.0)
    • Xcode at /Users/flutter/Downloads/Xcode14/Xcode.app/Contents/Developer
    • Build 14A309
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] Connected device (3 available)
    • iPhone 13 Pro (mobile) • D4909C3C-EA95-4F22-A52C-62D212410360 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-0 (simulator)
    • macOS (desktop)        • macos                                • darwin-x64     • macOS 12.6 21G115 darwin-x64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 113.0.5672.92

[✓] Network resources
    • All expected network resources are available.

• No issues found!

```

</details>

**Comment by EnduringBeta** [comment]
I can echo that what I'm experiencing seems to be the same issue after upgrading to Flutter 3.10. I have yet to see it visually, but it's appearing in integration tests using the iOS simulator. I use `AssetImage`.

**Comment by zrpra** [comment]
this also happened to me on iOS, it seems to be caused by the impeller engine as it also happens on android when i enable it. it only occurs on debug mode, on profile and release seems to be okay

**Comment by HasanAlqaisi** [comment]
I have the same issue on the IOS simulator. 
This happened after upgrading to v3.10.0.
I ran the app with `flutter run --no-enable-impeller` and it worked.

**Comment by dnfield** [comment]
I tried this out and could not reproduce it on a simulator.

I do notice that the doctor output is showing you're on Xcode 14.0. Can you try upgrading to the latest Xcode and see if it still reproduces? I'm running on Xcode 14.3.

**Comment by EnduringBeta** [comment]
> I have the same issue on the IOS simulator. This happened after upgrading to v3.10.0. I ran the app with `flutter run --no-enable-impeller` and it worked.

This worked for me and my GitHub Actions error when testing on iOS simulator.

**Comment by zrpra** [comment]
> @zrpra Does the issue still persist if you disable Impeller?

when impeller is disabled the image shows as normal

**Comment by sinahamedi67** [comment]
> @sinahamedi67 Since it's related to `picsum` API, the image could not be existing. So, is there any percentage occurrence of the issue? Or does it occur everytime?

Yes it occurs every time,  it happened on my own website first, all of the online images had the same problem. I used picsum just to write a sample.


> Does the issue still persist if you disable Impeller?

‌No, by disabling the impeller both my own project and sample works fine.

**Comment by huycozy** [comment]
@sinahamedi67 I also try on the same emulator as yours (iPhone 13 Pro, iOS 16.0) but still can not reproduce this. Can you try checking this on another emulator if any?

**Comment by Maieldd** [comment]
Meu smartphone tá sendo monitorado e eu acho que é por esse flutter ou sla, sei que tá conectado ao ios 16 e não sei o que fazer, tô triste demais

**Comment by msfstef** [comment]
we're experiencing the same issue, seems to only occur on CI but cannot reproduce locally - is there a workaround other than disabling Impeller?

**Comment by mmcdon20** [comment]
I am experiencing the same issue.

I made a github repo to reproduce this issue. 
(https://github.com/mmcdon20/displayimage/tree/master)

This error seems to happen on some iOS simulators but not others. 
For me specifically the iPhone 5s with iOS 12.4 is one simulator where this error occurs.

<img width="1792" alt="Screen Shot" src="https://github.com/flutter/flutter/assets/5104543/732c65d5-bdf4-4591-9859-8fd6423e44e1">

In the screenshot above both simulators are running the same app.

The simulator on the right is producing this error.

```

======== Exception caught by image resource service ================================================
The following _Exception was thrown resolving an image frame:
Exception: Codec failed to produce an image, possibly due to invalid image data.

When the exception was thrown, this was the stack: 
Image provider: AssetImage(bundle: null, name: "images/bottle.jpg")
Image key: AssetBundleImageKey(bundle: PlatformAssetBundle#5d83e(), name: "images/bottle.jpg", scale: 1.0)
====================================================================================================
```

The app runs on both if the `--no-enable-impeller` flag is passed, so appears to be directly related to impeller.

**Comment by dnfield** [comment]
Flutter does not support iPhone 5s anymore. Lowest version we support is 11.

**Comment by mmcdon20** [comment]
@dnfield iOS 11? The simulator is running iOS 12.4 on the 5s. You can see it in the screenshot that its 12.4.

**Comment by dnfield** [comment]
No, iPhone 11 :)

The lowest iPhone simulator I can create is an 8 locally. I can't even get runtimes at this point for lower than that. On an iPhone 8 simulator your demo app works fine for me locally.

**Comment by dnfield** [comment]
https://docs.flutter.dev/reference/supported-platforms#deploying-flutter

**Comment by mmcdon20** [comment]
@dnfield The table you are referencing appears to be iOS versions though, not iPhone versions (especially considering iPhone 16 doesn't exist yet). iOS 11-15 falls in the best effort category not the unsupported category.

**Comment by dnfield** [comment]
Ahh you are right.

Sorry about that. Let me check.

**Comment by dnfield** [comment]
I am unable to reproduce this on the oldest phone I can access. It still seems like this might be more of an issue on older simulators on older versions of macOS which I also don't currently have access to.

**Comment by Juliotati** [comment]
Just a heads up, I think this is probably not an impeller issue, because i'm experiencing the same issue on both android and iOS emulator and device.

```md
Flutter 3.7.12 • channel stable • https://github.com/flutter/flutter.git
Framework • revision 4d9e56e694 (4 weeks ago) • 2023-04-17 21:47:46 -0400
Engine • revision 1a65d409c7
Tools • Dart 2.19.6 • DevTools 2.20.1
```
![image](https://github.com/flutter/flutter/assets/59662912/3ceb9c9b-efa9-44e5-af14-7a1c47a35e3f)

**Comment by dnfield** [comment]
@Juliotati - you are seeing a similar exception message for a different reason. You are also on an older version of Flutter. AVIF support is platform specific in Flutter and not all platforms support it.

**Comment by sinahamedi67** [comment]
> @sinahamedi67 I also try on the same emulator as yours (iPhone 13 Pro, iOS 16.0) but still can not reproduce this. Can you try checking this on another emulator if any?

Yes it happens on other simulators too. Also I've upgraded my macos & xcode to the latest versions and still have this bug.
New flutter doctor:

``` 

[✓] Flutter (Channel stable, 3.10.0, on macOS 13.3.1 22E772610a darwin-x64, locale en-US)
    • Flutter version 3.10.0 on channel stable at /Users/flutter/Downloads/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 84a1e904f4 (9 days ago), 2023-05-09 07:41:44 -0700
    • Engine revision d44b5a94c9
    • Dart version 3.0.0
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at /Users/flutter/Library/Android/sdk
    • Platform android-33, build-tools 33.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3)
    • Xcode at /Users/flutter/Downloads/Xcode14.3/Xcode.app/Contents/Developer
    • Build 14E222b
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] Connected device (3 available)
    • iPhone 14 Pro (mobile) • 4B87186A-F646-430F-89EC-4A221B9C4CEB • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • macOS (desktop)        • macos                                • darwin-x64     • macOS 13.3.1 22E772610a darwin-x64
    • Chrome (web)           • chrome                               • web-javascript • Google Chrome 113.0.5672.126

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

**Comment by huycozy** [comment]
I still can not reproduce this issue on my end. The image is always shown as expected. If anyone is able to replicate this consistently and can find a pattern to replicate it, please write a comment.

**Comment by eduardofcr** [comment]
I have the same issues, for all pictures, with asset images, when using emulator.

**Comment by sinahamedi67** [comment]
Not sure if that has anything to do with it, but I'm using macOS on a VMware virtual machine. It seems that other users in this thread also have this bug in their CI/CD process. So perhaps Impeller relies on a library or dependency that doesn't function well on virtual machines.

**Comment by xonaman** [comment]
Same issue here, using iOS simulator

**Comment by fuzzybinary** [comment]
This is failing for me on iOS in CI as well, but not every time. I haven't been able to reproduce locally. The differences between environments:
* Processor architecture - Locally I have an M1 mac, remote is Intel
* Virtualization - Locally I'm not virtualizing, remotely I believe our CI provider is
* Hardware acceleration - Locally I have hardware acceleration, CI is headless.

Both should be using iPhone 14 simulators on the latest versions of XCode and iOS.

Going go work around by using `--no-enable-impeller` on CI for now.

**Comment by chinmaygarde** [comment]
@sinahamedi67 @fuzzybinary The observation that this happens on CI that may have a virtualized environment is intriguing. And could also explain our difficulty in reproducing it.

W.r.t texture management, Impeller differs from the existing backend in one key way. Before Impeller, when decompressing images, the engine would create an uncompressed texture onto the heap before transferring this to GPU device memory and further modifying its usage there. Impeller elides this copy on the heap and directly decompresses into GPU device memory mapped into the process. This reduces memory usage spikes during texture decompression. But this also depends on carefully considering device capabilities. We thought we did our due diligence by testing on the devices and simulators but haven't considered the implications of a virtualized environment with GPU passthrough.

We still haven't reproduced this issue ourselves but we are speculatively disabling this optimization on the simulators in https://github.com/flutter/engine/pull/42161 while we investigate some more.

Some potential threads to keep pulling on:
* @jonahwilliams @dnfield Can we land and cherry-pick the speculative "fix" by Monday? Its low risk and simulator only. We can then get more feedback. With luck, folks won't have to work around this like @fuzzybinary had to.
* We should prioritize https://github.com/flutter/flutter/issues/127061 ASAP so the error message is less vague. At least we'll know where the failure is (I suspect its in the call to map GPU memory in the process). My guess is that the patch will be large enough to trip our process checks for cherry-pick viability. But maybe not.
* @sinahamedi67 @fuzzybinary If possible, can you give us details on the kind of virtualization used by your CI providers? @sinahamedi67 mentioned VMWare. @fuzzybinary Are you using the same? Our corp. boxes have restrictions on virtualization software so we'll have to reproduce in a clean room environment anyway.
* Seeking proof by contradiction, can anyone on this thread reproduce this in a known non-virtualized environment? If so, can you paste the information in the "About this Mac" and "System Information -> Graphics/Displays" sections?
* I'll follow up on [clarifying support for virtualized environments](https://github.com/flutter/flutter/issues/127210) and see how we can more easily debug such issues now and in the future.

**Comment by dnfield** [comment]
> * @jonahwilliams @dnfield Can we land and cherry-pick the speculative "fix" by Monday? Its low risk and simulator only. We can then get more feedback. With luck, folks won't have to work around this like @fuzzybinary had to.

I think it's a low risk cherry pick, but once it rolls into the framework it' dbe great if someone on this thread who's been able to reproduce the issue (e.g. on CI) can make sure the issue goes away with that change. Otherwise we're cherry picking a pretty speculative fix that might cause other issues and has no tests...

**Comment by jonahwilliams** [comment]
I would add that unless someone can confirm that the patch works (once it rolls into main) we will not cherry pick it.

**Comment by jonahwilliams** [comment]
The flutter commit d211dc141a98d21bb5ebb2baddc3cd4cc9e64f27 contains an updated version of image decoding for the simulators. Someone who can reproduce the issue will need to verify the change had an impact. We're also working on bubbling up the exact error message from the engine.

**Comment by sinahamedi67** [comment]
> The flutter commit [d211dc1](https://github.com/flutter/flutter/commit/d211dc141a98d21bb5ebb2baddc3cd4cc9e64f27) contains an updated version of image decoding for the simulators. Someone who can reproduce the issue will need to verify the change had an impact. We're also working on bubbling up the exact error message from the engine.

I checked again with latest version `Channel master, 3.11.0-10.0.pre.44` which had that commit, but didn't change the result.

**Comment by baconcheese113** [comment]
This is happening on my test runs, not sure how to disable impeller using `flutter test`

https://github.com/baconcheese113/handle-it-app/actions/runs/5035266919/jobs/9030659451#step:14:18380

**Comment by jonahwilliams** [comment]
Flutter test does not use impeller

**Comment by baconcheese113** [comment]
I no longer get the error in this thread (`Codec failed to produce an image`) when I downgrade
https://github.com/baconcheese113/handle-it-app/actions/runs/5041544099/jobs/9041331691

Downgrading from flutter 3.10.0 back to 3.7.7, dart 3.0.0 to 2.17.4, and collection 1.17.1 to 1.17.0

I tested with the iPhone 11 and iPhone 12 simulators

**Comment by sdegroot** [comment]
> * Seeking proof by contradiction, can anyone on this thread reproduce this in a known non-virtualized environment? If so, can you paste the information in the "About this Mac" and "System Information -> Graphics/Displays" sections?

I have this issue and am not working in a virtualised environment. 

```
Apple M1 Pro:

  Chipset Model:	Apple M1 Pro
  Type:	GPU
  Bus:	Built-In
  Total Number of Cores:	16
  Vendor:	Apple (0x106b)
  Metal Support:	Metal 3
  Displays:
Colour LCD:
  Display Type:	Built-in Liquid Retina XDR Display
  Resolution:	3024x1964 Retina
  Main Display:	Yes
  Mirror:	Off
  Online:	Yes
  Automatically Adjust Brightness:	Yes
  Connection Type:	Internal
C34H89x:
  Resolution:	3440x1440 (UWQHD - Ultra-Wide Quad HD)
  UI Looks like:	3440 x 1440 @ 100.00Hz
  Mirror:	Off
  Online:	Yes
  Rotation:	Supported
```

Leading to 

```
======== Exception caught by image resource service ================================================
The following _Exception was thrown resolving an image frame:
Exception: Codec failed to produce an image, possibly due to invalid image data.

When the exception was thrown, this was the stack: 
Image provider: AssetImage(bundle: null, name: "assets/logo/circular.png")
Image key: AssetBundleImageKey(bundle: PlatformAssetBundle#686e0(), name: "assets/logo/circular.png", scale: 1.0)
```

How can I help?

**Comment by jason-simmons** [comment]
Reproduced this in the iOS simulator within a VM running macOS Monterey.

In that environment `ShouldUseMetalRenderer` is returning false because the Metal device does not support the desired feature set.  `GetRenderingAPIForProcess` then selects the Skia-based software renderer instead of the Impeller Metal renderer.

When an image is decoded `ImageDecoder::Make` selects the `ImageDecoderImpeller` based on the `settings.enable_impeller` flag.  But if the engine is using the software renderer, then the `ImageDecoderImpeller` can not obtain an Impeller context and therefore fails.

**Comment by chinmaygarde** [comment]
**TL;DR: Metal needs to be enabled on a virtualized macOS environment for Flutter simulators.**

We discussed Jason's findings as a team and tried to reason about the observed behavior as it was extremely surprising to us.

Our understanding as of right now is as follows. Please [disregard our initial hunch as that didn’t pan out](https://github.com/flutter/flutter/issues/126768#issuecomment-1554945039) and turned out to be the wrong line of investigation.



* Jason’s findings revealed that Flutter could not access the Metal device to set up rendering. It then fell back to an unsupported software rendering configuration.
* [Flutters system requirements](https://docs.flutter.dev/get-started/install/macos) for macOS hosts specifies a 10.14 (Mojave) version floor. This version was selected because it was the first version that started [requiring a GPU capable of supporting Metal](https://en.wikipedia.org/wiki/MacOS_Mojave#:~:text=Mojave%20requires%20a%20GPU%20that%20supports%20Metal).
* CI solutions we were aware of including the one used by Flutter itself use bare metal (i.e, non-virtualized) environments. Because of the version requirement, there was not an environment Flutter supported (simulator or otherwise) that didn’t also require Metal support.
* Investigating how this worked at all so far, we found an old code path that wasn’t cleaned up. It used to be that Flutter supported the OpenGL backend. But Apple always emulated OpenGL in software and developers complained that the OpenGL backend was too slow. So we enabled a direct software backend.
* This software backend was also too slow and we finally added a Metal simulator backend once Mojave became a requirement. But when we did, we didn’t remove the software fallback. There is this goofy [if-this-else-this](https://github.com/flutter/engine/blob/8b6617e3f36fc45a122ffb0b90aab23a196483a9/shell/platform/darwin/ios/rendering_api_selection.mm#L50) relic in version selection today for those interested. There was no negative test.
* So virtualized CI environments that don’t respect the GPU requirements of Mojave and later have been using a software configuration that is not supported and hasn’t been for maybe a couple of years.
* Investigating GPU passthrough support in virtualized environments, both [VMWare Fusion](https://kb.vmware.com/s/article/81657) and [Parallels](https://www.parallels.com/blogs/parallels-desktop-15-directx-11-and-metal/) seem to support it but may require configuration. Parallels documents it as being available by default. Other guidance on the internet seems to [list compatibility issues](https://pspdfkit.com/blog/2020/managing-macos-hardware-virtualization-or-bare-metal/#compatibility). Unfortunately, our experience with virtualized hosts and debugging issues with them is limited.
* Looking into @fuzzybinary s CI provider, [I found this guidance on their blog saying Metal should be supported](https://bitrise.io/blog/post/run-faster-ios-ui-tests-via-metal-support). But I see a retraction later that they were pausing support for “a short period of time”. That retraction is over a year old so we could not tell if Bitrise supports Metal or not. Or if there is an opt-in.
* Since Flutter’s system requirements specify a macOS version that requires Metal, we are inclined to terminate Flutter engine launch (with a clear message) if that requirement is not met. The workaround to not use Impeller on such hosts may be preferable but the renderer will be quite dissimilar to rendering on actual devices. Once the legacy backend is removed, this will stop working as well. The real fix would be to enable Metal support on virtualized environments.

I’ll follow up on trying to figure out the impact of this as I’m not sure how many CI providers that use virtualization just flat out refuse Metal device creation.

**Comment by chinmaygarde** [comment]
Doing some more investigating on the impact of this on CI environments:

* The big one seems to be the GitHub actions. But GPU passthrough was apparently added with the macos-13 image. Investigating @baconcheese113 s [linked issue](https://github.com/flutter/flutter/issues/126768#issuecomment-1556330720), it seems like [the workflow targets](https://github.com/baconcheese113/handle-it-app/actions/runs/5035266919/workflow#L26) "macos-latest". Took me a while to figure out that "latest" is still on 12 as the status on the [rollout of 13 is not complete](https://github.com/actions/runner-images#available-images). Specifying it explicitly should fix issues with passthrough on GitHub actions. Apparently, this issue has affected macOS desktop CI configurations and those don't have a fallback at all. I'll cross the issue here when I find it.
* Both VMWare and Parallels have GPU passthrough support. Parallels does this by default too. I've linked the resources I could find on the matter in my update above.
* I couldn't find a clear indication on the status of GPU passthrough with Bitrise (used by @fuzzybinary). The messaging on their blog was confusing but it seems like it should at least be possible.
* @sdegroot s update indicates that he could not reproduce this in a non-virtualized environment. I'll keep an eye out in any case.

Since CI environments should support GPU passthrough. We are inclined to make the non-availability of [Metal on simulators a fatal error](https://github.com/flutter/flutter/issues/127408) since we weren't even aware folks were routinely running into this unsupported configuration.

Thanks everyone for help pinpointing this. Closing this as there are no pending action items.

**Comment by chinmaygarde** [comment]
xref [the macOS Desktop CI issue](https://github.com/flutter/flutter/issues/118469).
